File: go/1622.go
package main

const FancyMod = 1000000007

func pow(base, exp int) int {
	res := 1
	base %= FancyMod
	for exp > 0 {
		if exp%2 == 1 {
			res = (res * base) % FancyMod
		}

		base = (base * base) % FancyMod
		exp /= 2
	}
	return res
}

type Fancy struct {
	sequence []int
	mult     int
	inc      int
}

func Constructor() Fancy {
	return Fancy{
		sequence: []int{},
		mult:     1,
		inc:      0,
	}
}

func (this *Fancy) Append(val int) {
	// (val - add) * inv(mul)
	temp := (val - this.inc + FancyMod) % FancyMod
	this.sequence = append(this.sequence, (temp*pow(this.mult, FancyMod-2))%FancyMod)
}

func (this *Fancy) AddAll(inc int) {
	this.inc = (this.inc + inc) % FancyMod
}

func (this *Fancy) MultAll(m int) {
	this.mult = (this.mult * m) % FancyMod
	this.inc = (this.inc * m) % FancyMod // need to distribute the multiplier
}

func (this *Fancy) GetIndex(idx int) int {
	if idx >= len(this.sequence) {
		return -1
	}

	res := (this.sequence[idx] * this.mult) % FancyMod
	res = (res + this.inc) % FancyMod

	return res
}

File: zig/1622.zig
const std = @import("std");

const MOD = 1000000007;

const FancyValue = struct {
    val: i32,
    idx: usize,
};

const FancyOperation = struct {
    mult: bool,
    val: i32,
};

const Fancy = struct {
    sequence: std.ArrayList(FancyValue),
    operations: std.ArrayList(FancyOperation),

    pub fn construct() Fancy {
        return Fancy{ .sequence = .empty, .operations = .empty };
    }

    pub fn append(self: *Fancy, allocator: std.mem.Allocator, val: i32) void {
        try self.sequence.append(allocator, FancyValue{ .val = val, .idx = self.operations.items.len });
    }

    pub fn addAll(self: *Fancy, allocator: std.mem.Allocator, inc: i32) void {
        try self.operations.append(allocator, FancyOperation{ .mult = false, .val = inc });
    }

    pub fn multAll(self: *Fancy, allocator: std.mem.Allocator, m: i32) void {
        try self.operations.append(allocator, FancyOperation{ .mult = false, .val = m });
    }

    pub fn getIndex(self: *Fancy, idx: usize) i32 {
        if (idx >= self.sequence.items.len) {
            return -1;
        }

        for (self.sequence.items[idx].idx..self.operations.items.len) |i| {
            if (self.operations.items[i].mult) {
                self.sequence.items[idx].val = @mod(self.sequence.items[idx].val * self.operations.items[i].val, MOD);
            } else {
                self.sequence.items[idx].val = @mod(self.sequence.items[idx].val + self.operations.items[i].val, MOD);
            }
        }

        self.sequence.items[idx].idx = self.operations.items.len;

        return self.sequence.items[idx].val;
    }
};
